pkg/apis/alicloud/validation/infrastructure.go

```diff
@@ -97,8 +100,50 @@ func ValidateInfrastructureConfig(infra *apisalicloud.InfrastructureConfig, node
 func ValidateInfrastructureConfigUpdate(oldConfig, newConfig *apisalicloud.InfrastructureConfig) field.ErrorList {
 	allErrs := field.ErrorList{}
 
+	var (
+		newZonesWithoutNatGatewayConfig = make([]apisalicloud.Zone, 0, len(newConfig.Networks.Zones))
+		oldZonesWithoutNatGatewayConfig = make([]apisalicloud.Zone, 0, len(oldConfig.Networks.Zones))
+
+		networksPath = field.NewPath("networks")
+	)
+
+	for i, zone := range newConfig.Networks.Zones {
+		newZonesWithoutNatGatewayConfig = append(newZonesWithoutNatGatewayConfig, apisalicloud.Zone{
+			Name:    zone.Name,
+			Worker:  zone.Worker,
+			Workers: zone.Workers,
+		})
+
+		allErrs = append(allErrs, ValidateNatGatewayConfig(zone.NatGateway, networksPath.Child("zones").Index(i).Child("natGateway").Child("ipAddresses"))...)
+	}
+
+	for i, zone := range oldConfig.Networks.Zones {
+		oldZonesWithoutNatGatewayConfig = append(oldZonesWithoutNatGatewayConfig, apisalicloud.Zone{
+			Name:    zone.Name,
+			Worker:  zone.Worker,
+			Workers: zone.Workers,
+		})
+
+		allErrs = append(allErrs, ValidateNatGatewayConfig(zone.NatGateway, networksPath.Child("zones").Index(i).Child("natGateway").Child("ipAddresses"))...)
+	}
+
 	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newConfig.Networks.VPC, oldConfig.Networks.VPC, field.NewPath("networks").Child("vpc"))...)
-	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newConfig.Networks.Zones, oldConfig.Networks.Zones, field.NewPath("networks").Child("zones"))...)
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(newZonesWithoutNatGatewayConfig, oldZonesWithoutNatGatewayConfig, field.NewPath("networks").Child("zones"))...)
```

Comment:
TBH, I don't really understand this code. What is happening here? Is there a unit test for it?

Reply:
The code is to allow user update the field `natGateway` under the `zones`, and they have the follow structure.
```
networks:
  vpc:
    cidr: 10.250.0.0/16
  zones:
  - name: cn-shanghai-e
    worker: 10.250.0.0/19
    natGateway:
      ipAddresses: 
```
Because of this code `allErrs = append(allErrs, apivalidation.ValidateImmutableField(newConfig.Networks.Zones, oldConfig.Networks.Zones, field.NewPath("networks").Child("zones"))...)`, it will forbid the update of all fields under `zones`. In order to make `natGateway` could be updated and other fields not allowed, I generate two new `Zones` structure without `natGateway` info to validate if there is some change or not on other fields. 

If I have some misunderstanding of the code logic, please correct me. :)

Reply:
Ah, interesting 😄 I see, understood now. Maybe a comment would be helpful for readers.